x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/ssl/SSLReloadDuringStartupIntegTests.java

```diff
@@ -59,66 +77,123 @@ protected boolean transportSSLEnabled() {
-                        writtenLatch.countDown();
+                        atomicCopyIfPossible(goodKeystorePath, nodeKeystorePath);
+                        afterKeystoreFix.countDown(); // Indicate keystore is fixed
                     } catch (IOException e) {
                         throw new UncheckedIOException(e);
                     }
-                }).start();
-
-                latch.countDown();
-                try {
-                    latch.await();
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                }
-                return super.onNodeStopped(nodeName);
+                });
+                fixKeystoreThread.setName("Fix Keystore");
+                fixKeystoreThread.start();
+                waitUntilFixKeystoreIsReadyToBegin(beforeKeystoreFix); // SYNCHRONIZED TASK
+                return super.onNodeStopped(nodeName); // Return control to RestartCallback to do startup next
             }
         });
-        writtenLatch.await();
-        ensureClusterSizeConsistency();
-        ensureFullyConnectedCluster();
+        final long awaitNanos = System.nanoTime(); // actual await nanos
+        try {
+            afterKeystoreFix.await(); // Wait until keystore is fixed
+        } finally {
+            LOGGER.info("Awaited {}ms. Verifying the cluster...", (System.nanoTime() - awaitNanos) / 1000000F);
+        }
+        final long ensureClusterSizeConsistencyNanos = System.nanoTime(); // actual verify nanos
+        try {
+            ensureClusterSizeConsistency();
+        } finally {
+            LOGGER.info("Ensure cluster size consistency took {}ms.", (System.nanoTime() - ensureClusterSizeConsistencyNanos) / 1000000F);
+        }
+        final long ensureFullyConnectedClusterNanos = System.nanoTime(); // actual verify nanos
+        try {
+            ensureFullyConnectedCluster();
+        } finally {
+            LOGGER.info("Ensure fully connected cluster took {}ms.", (System.nanoTime() - ensureFullyConnectedClusterNanos) / 1000000F);
+        }
+    }
+
+    private void waitUntilNodeStartupIsReadyToBegin(final CountDownLatch beforeKeystoreFix) {
+        beforeKeystoreFix.countDown(); // Fixer thread ready
+        try {
+            final long sleepMillis = randomLongBetween(200L, 600L); // intended sleepMillis
+            final long awaitNanos = System.nanoTime(); // actual await nanos
+            try {
+                beforeKeystoreFix.await(); // Wait for Main thread to be ready for concurrent start
+            } finally {
+                LOGGER.info("Awaited {}ms. Sleeping {}ms before fixing...", (System.nanoTime() - awaitNanos) / 1000000F, sleepMillis);
+            }
+            long sleepNanos = System.nanoTime(); // actual sleep nanos
+            try {
+                Thread.sleep(sleepMillis); // Note: nodeSettings() => `resource.reload.interval.high: 1s`
```

Comment:
We typically try to avoid putting sleep in the tests if at all possible. It slows down the testing and can make troubleshooting more difficult. It is generally preferable to wait or busy wait (with an optional timeout) on some state.

Reply:
I agree the random sleep is not ideal, but my intention is not to change the functionality. Sleep was already there, and it may still be needed to reproduce the issue.

For context, this thread in the original PR explains why the latch (and sleep) are there.
https://github.com/elastic/elasticsearch/pull/54999#discussion_r409241976

I considered updating RestartCallback to add onStarting() and onStarted() hooks. The test could use both for additional use logs, and onStarting() could replace sleep. However, I think that refactoring should be deferred to a different PR.

What do you think?

Reply:
I missed it was already there. I agree not changing functionality is best.